File: backend/data/skill_store/imported/external/bio-single-cell-data-io/scripts/python/samplesheet.py
"""SampleSheet-based single-cell data loading for Scanpy.

Reference: scanpy 1.10+, anndata 0.10+, pandas 2.2+

This module provides:
- read_samplesheet(): validate and read a SampleSheet CSV
- load_from_samplesheet(): load all samples and optionally merge

SampleSheet format (CSV):
    sample_id,file_path,file_format,condition,batch,technology
    PA08,data/PA08/filtered_feature_bc_matrix,10x_mtx,High_NI,Batch1,10x_v3
    PA11,data/PA11/filtered_feature_bc_matrix,10x_mtx,High_NI,Batch1,10x_v3

Required columns: sample_id, file_path, file_format
Optional columns: condition, batch, technology, sex, age, note
"""

from typing import List, Union

import anndata
import pandas as pd
import scanpy as sc

from geo_loaders import load_geo_h5, load_geo_mtx_with_metadata
from utils import SUPPORTED_FORMATS, strip_barcode_suffix, validate_file_exists


def read_samplesheet(sheet_path: str) -> pd.DataFrame:
    """Read and validate a SampleSheet CSV.

    Validation rules:
    - Required columns: sample_id, file_path, file_format
    - sample_id must be unique and non-empty
    - file_path must exist
    - file_format must be in SUPPORTED_FORMATS
    - Warns if batch column missing when n_samples > 1

    Parameters
    ----------
    sheet_path : str
        Path to SampleSheet CSV.

    Returns
    -------
    Validated DataFrame with one row per sample.
    """
    validate_file_exists(sheet_path, "SampleSheet")

    sheet = pd.read_csv(sheet_path)

    # Required columns
    required = {"sample_id", "file_path", "file_format"}
    missing = required - set(sheet.columns)
    if missing:
        raise ValueError(f"SampleSheet missing required columns: {missing}")

    # Unique sample_id
    if sheet["sample_id"].duplicated().any():
        dups = sheet["sample_id"][sheet["sample_id"].duplicated()].unique().tolist()
        raise ValueError(f"Duplicate sample_id found: {dups}")

    # Non-empty sample_id
    empty_ids = sheet["sample_id"].isna() | (sheet["sample_id"].astype(str).str.strip() == "")
    if empty_ids.any():
        raise ValueError("sample_id contains empty or NA values")

    # File existence
    for _, row in sheet.iterrows():
        if not pd.isna(row["file_path"]):
            validate_file_exists(str(row["file_path"]), f"sample '{row['sample_id']}'")

    # Format validation
    invalid_formats = set(sheet["file_format"].dropna()) - set(SUPPORTED_FORMATS)
    if invalid_formats:
        raise ValueError(
            f"Unsupported file_format: {invalid_formats}. "
            f"Supported: {SUPPORTED_FORMATS}"
        )

    # Warn if batch missing with >1 sample
    if len(sheet) > 1 and "batch" not in sheet.columns:
        print(
            "WARNING: Multiple samples detected but 'batch' column not found. "
            "Consider adding batch info for downstream integration."
        )

    return sheet


def load_from_samplesheet(
    sheet_path: str, merge: bool = True, join: str = "outer"
) -> Union[sc.AnnData, List[sc.AnnData]]:
    """Load all samples from a SampleSheet.

    Parameters
    ----------
    sheet_path : str
        Path to SampleSheet CSV.
    merge : bool, default True
        If True and n_samples > 1, concatenate into single AnnData.
        If False, return list of AnnData objects.
    join : str, default "outer"
        Join strategy for concatenation ("outer" or "inner").

    Returns
    -------
    AnnData if merge=True, List[AnnData] if merge=False.

    Notes
    -----
    - Each sample's metadata (condition, batch, technology, etc.) is injected
      into adata.obs.
    - For geo_mtx_merged format, the metadata_csv path is expected to be in
      the same directory as the MTX, named ``<sample_id>_metadata.csv`` or
      specified via a ``metadata_csv`` column in the SampleSheet.
    - Barcode suffixes (e.g. "-1") are automatically stripped when aligning
      merged MTX with metadata.
    """
    sheet = read_samplesheet(sheet_path)

    adata_list = []
    for _, row in sheet.iterrows():
        sample_id = str(row["sample_id"])
        file_path = str(row["file_path"])
        fmt = str(row["file_format"])

        # Route to format-specific loader
        adata = _load_by_format(file_path, fmt, sample_id, row)

        # Inject metadata from SampleSheet optional columns
        for col in ["condition", "batch", "technology", "sex", "age", "note"]:
            if col in row and pd.notna(row[col]):
                adata.obs[col] = row[col]

        adata_list.append(adata)

    # Named list by sample_id
    names = sheet["sample_id"].astype(str).tolist()
    for ad, nm in zip(adata_list, names):
        ad.obs["sample_id"] = nm

    # Single sample: return directly regardless of merge flag
    if len(adata_list) == 1:
        return adata_list[0]

    if not merge:
        return adata_list

    # Concatenate
    adata_merged = anndata.concat(
        adata_list,
        join=join,
        label="sample_id",
        keys=names,
        index_unique="-",
    )

    # Ensure categorical types
    adata_merged.obs["sample_id"] = adata_merged.obs["sample_id"].astype("category")
    for col in ["batch", "condition", "technology"]:
        if col in adata_merged.obs.columns:
            adata_merged.obs[col] = adata_merged.obs[col].astype("category")

    print(
        f"Merged {len(adata_list)} samples: "
        f"{adata_merged.n_obs} cells x {adata_merged.n_vars} genes"
    )

    return adata_merged


def _load_by_format(
    file_path: str, fmt: str, sample_id: str, sheet_row: pd.Series
) -> sc.AnnData:
    """Dispatch to the correct loader based on file_format."""
    if fmt == "10x_mtx":
        adata = sc.read_10x_mtx(file_path, var_names="gene_symbols", cache=False)
        adata.var_names_make_unique()
        return adata

    if fmt == "10x_h5":
        adata = sc.read_10x_h5(file_path)
        adata.var_names_make_unique()
        return adata

    if fmt == "geo_mtx":
        adata = sc.read_10x_mtx(file_path, var_names="gene_symbols", cache=False)
        adata.var_names_make_unique()
        return adata

    if fmt == "geo_mtx_merged":
        # metadata_csv can be specified in SampleSheet or inferred
        metadata_csv = None
        import os

        base = os.path.dirname(file_path)
        if "metadata_csv" in sheet_row and pd.notna(sheet_row["metadata_csv"]):
            metadata_csv = str(sheet_row["metadata_csv"])
            if not os.path.exists(metadata_csv):
                raise FileNotFoundError(
                    f"SampleSheet specifies metadata_csv='{metadata_csv}' but file not found. "
                    f"Check the path or place metadata.csv in {base}"
                )
        else:
            # Try common naming patterns
            candidates = [
                os.path.join(base, f"{sample_id}_metadata.csv"),
                os.path.join(base, "metadata.csv"),
                os.path.join(base, "cell_metadata.csv"),
            ]
            for cand in candidates:
                if os.path.exists(cand):
                    metadata_csv = cand
                    break
        if metadata_csv is None:
            raise ValueError(
                f"geo_mtx_merged requires metadata_csv. "
                f"Add 'metadata_csv' column to SampleSheet or place "
                f"metadata.csv in {base}"
            )
        return load_geo_mtx_with_metadata(
            mtx_dir=file_path,
            metadata_csv=metadata_csv,
            sample_col="sample_id" if "sample_id" in pd.read_csv(metadata_csv, nrows=0).columns else "sample",
        )

    if fmt == "geo_h5":
        return load_geo_h5(file_path, sample_id=sample_id)

    if fmt == "h5ad":
        adata = sc.read_h5ad(file_path)
        if "sample_id" not in adata.obs.columns:
            adata.obs["sample_id"] = sample_id
        return adata

    if fmt == "rds":
        raise ValueError(
            f"Cannot load RDS file '{file_path}' in Python. "
            f"Use R workflow or convert to h5ad first."
        )

    raise ValueError(f"Unsupported format: {fmt}")

File: backend/data/skill_store/imported/nanoresearch/bio-single-cell-data-io/scripts/r/samplesheet.R
#' SampleSheet-based single-cell data loading for Seurat
#'
#' Reference: Seurat 5.0+
#'
#' This module provides:
#' - read_samplesheet(): validate and read a SampleSheet CSV
#' - load_from_samplesheet(): load all samples and optionally merge
#'
#' SampleSheet format (CSV):
#'     sample_id,file_path,file_format,condition,batch,technology
#'     PA08,data/PA08/filtered_feature_bc_matrix,10x_mtx,High_NI,Batch1,10x_v3
#'     PA11,data/PA11/filtered_feature_bc_matrix,10x_mtx,High_NI,Batch1,10x_v3
#'
#' Required columns: sample_id, file_path, file_format
#' Optional columns: condition, batch, technology, sex, age, note

library(Seurat)


# ---------------------------------------------------------------------------
# SampleSheet I/O
# ---------------------------------------------------------------------------

#' Read and validate a SampleSheet CSV
#'
#' Validation rules:
#' - Required columns: sample_id, file_path, file_format
#' - sample_id must be unique and non-empty
#' - file_path must exist
#' - file_format must be in SUPPORTED_FORMATS
#' - Warns if batch column missing when n_samples > 1
#'
#' @param sheet_path Path to SampleSheet CSV.
#' @return Validated data.frame with one row per sample.
#' @export
read_samplesheet <- function(sheet_path) {
  if (!file.exists(sheet_path)) {
    stop("SampleSheet not found: ", sheet_path)
  }

  sheet <- read.csv(sheet_path, stringsAsFactors = FALSE, check.names = FALSE)

  # Required columns
  required <- c("sample_id", "file_path", "file_format")
  missing <- setdiff(required, colnames(sheet))
  if (length(missing) > 0) {
    stop("SampleSheet missing required columns: ", paste(missing, collapse = ", "))
  }

  # Unique sample_id
  if (any(duplicated(sheet$sample_id))) {
    dups <- unique(sheet$sample_id[duplicated(sheet$sample_id)])
    stop("Duplicate sample_id found: ", paste(dups, collapse = ", "))
  }

  # Non-empty sample_id
  empty_ids <- is.na(sheet$sample_id) | nzchar(trimws(sheet$sample_id)) == FALSE
  if (any(empty_ids)) {
    stop("sample_id contains empty or NA values")
  }

  # File existence
  for (i in seq_len(nrow(sheet))) {
    if (!file.exists(sheet$file_path[i])) {
      stop(
        "Path not found for sample '", sheet$sample_id[i], "': ",
        sheet$file_path[i]
      )
    }
  }

  # Format validation
  invalid <- setdiff(sheet$file_format, SUPPORTED_FORMATS)
  if (length(invalid) > 0) {
    stop(
      "Unsupported file_format: ", paste(invalid, collapse = ", "),
      ". Supported: ", paste(SUPPORTED_FORMATS, collapse = ", ")
    )
  }

  # Warn if batch missing with >1 sample
  if (nrow(sheet) > 1 && !("batch" %in% colnames(sheet))) {
    warning(
      "Multiple samples detected but 'batch' column not found. ",
      "Consider adding batch info for downstream integration.",
      call. = FALSE
    )
  }

  return(sheet)
}


# ---------------------------------------------------------------------------
# Main loader
# ---------------------------------------------------------------------------

#' Load all samples from a SampleSheet
#'
#' @param sheet_path Path to SampleSheet CSV.
#' @param merge Logical. If TRUE (default), merge all samples into one
#'   Seurat object. If FALSE, return a named list of Seurat objects.
#' @param min.cells,min.features Passed to CreateSeuratObject.
#' @return Seurat object if merge=TRUE, named list if merge=FALSE.
#'   For a single sample, always returns a Seurat object.
#' @export
load_from_samplesheet <- function(
  sheet_path,
  merge = TRUE,
  min.cells = 3,
  min.features = 200
) {
  sheet <- read_samplesheet(sheet_path)

  obj_list <- list()

  for (i in seq_len(nrow(sheet))) {
    row <- sheet[i, ]
    sample_id <- row$sample_id
    file_path <- row$file_path
    fmt <- row$file_format

    # Route to format-specific loader
    obj <- .load_one_sample(
      file_path = file_path,
      fmt = fmt,
      sample_id = sample_id,
      sheet_row = row,
      min.cells = min.cells,
      min.features = min.features
    )

    # Inject metadata from SampleSheet optional columns
    for (col in c("condition", "batch", "technology", "sex", "age", "note")) {
      if (col %in% colnames(row) && !is.na(row[[col]])) {
        obj[[col]] <- row[[col]]
      }
    }

    obj_list[[sample_id]] <- obj
  }

  # Single sample: return directly
  if (length(obj_list) == 1) {
    return(obj_list[[1]])
  }

  if (!merge) {
    return(obj_list)
  }

  # Merge with add.cell.ids
  merged <- merge(
    obj_list[[1]],
    y = obj_list[-1],
    add.cell.ids = names(obj_list)
  )

  message(sprintf(
    "Merged %d samples: %d cells x %d features",
    length(obj_list), ncol(merged), nrow(merged)
  ))

  return(merged)
}


# ---------------------------------------------------------------------------
# Format-specific loaders
# ---------------------------------------------------------------------------

.load_one_sample <- function(
  file_path, fmt, sample_id, sheet_row,
  min.cells = 3, min.features = 200
) {
  if (fmt == "10x_mtx") {
    counts <- Read10X(data.dir = file_path)
    obj <- CreateSeuratObject(
      counts = counts, project = sample_id,
      min.cells = min.cells, min.features = min.features
    )
    obj$sample_id <- sample_id
    return(obj)
  }

  if (fmt == "10x_h5") {
    counts <- Read10X_h5(filename = file_path)
    obj <- CreateSeuratObject(
      counts = counts, project = sample_id,
      min.cells = min.cells, min.features = min.features
    )
    obj$sample_id <- sample_id
    return(obj)
  }

  if (fmt == "geo_mtx") {
    counts <- Read10X(data.dir = file_path)
    obj <- CreateSeuratObject(
      counts = counts, project = sample_id,
      min.cells = min.cells, min.features = min.features
    )
    obj$sample_id <- sample_id
    return(obj)
  }

  if (fmt == "geo_mtx_merged") {
    metadata_csv <- NULL
    if ("metadata_csv" %in% colnames(sheet_row) && !is.na(sheet_row$metadata_csv)) {
      metadata_csv <- sheet_row$metadata_csv
    } else {
      candidates <- c(
        file.path(dirname(file_path), paste0(sample_id, "_metadata.csv")),
        file.path(dirname(file_path), "metadata.csv"),
        file.path(dirname(file_path), "cell_metadata.csv")
      )
      for (cand in candidates) {
        if (file.exists(cand)) {
          metadata_csv <- cand
          break
        }
      }
    }
    if (is.null(metadata_csv)) {
      stop(
        "geo_mtx_merged requires metadata_csv. ",
        "Add 'metadata_csv' column to SampleSheet or place ",
        "metadata.csv in ", dirname(file_path)
      )
    }
    return(load_geo_mtx_merged(
      mtx_dir = file_path,
      metadata_csv = metadata_csv,
      sample_col = "sample_id",
      min.cells = min.cells,
      min.features = min.features
    ))
  }

  if (fmt == "geo_h5") {
    return(load_geo_h5(
      h5_path = file_path,
      sample_id = sample_id,
      min.cells = min.cells,
      min.features = min.features
    ))
  }

  if (fmt == "rds") {
    obj <- readRDS(file_path)
    if (!inherits(obj, "Seurat")) {
      stop("RDS file does not contain a Seurat object: ", file_path)
    }
    if (!("sample_id" %in% colnames(obj@meta.data))) {
      obj$sample_id <- sample_id
    }
    return(obj)
  }

  if (fmt == "h5ad") {
    stop(
      "Cannot load h5ad file '", file_path, "' in R. ",
      "Use Python workflow or convert to RDS first."
    )
  }

  stop("Unsupported format: ", fmt)
}
